app/src/org/commcare/appupdate/CommcareFlexibleAppUpdateManager.java

```diff
@@ -0,0 +1,206 @@
+    private final Runnable mCallback;
+    private boolean mEnabled;
+
+    private @UpdateAvailability int mUpdateAvailability;
+    private @InstallStatus int mInstallStatus;
+    private AppUpdateState mAppUpdateState;
+
+    private int percentageDownloaded;
+    private @InstallErrorCode int mInstallErrorCode = InstallErrorCode.NO_ERROR;
+
+    /**
+     * callback {@link Runnable} to notify when update state changes.
+     */
+    CommcareFlexibleAppUpdateManager(Runnable callback, AppUpdateManager appUpdateManager) {
+        this.mCallback = callback;
+        this.mAppUpdateManager = appUpdateManager;
+        mAppUpdateState = null;
+    }
+
+    //region AppUpdateController implementation
+    @Override
+    public void register() {
+        mEnabled = true;
+        mAppUpdateManager.registerListener(this);
+        fetchAppUpdateInfo();
+    }
+
+    @Override
+    public void unregister() {
+        mEnabled = false;
+        mAppUpdateManager.unregisterListener(this);
+    }
+
+    @Override
+    public void startUpdate(Activity activity) {
+        try {
+            boolean success = mAppUpdateManager.startUpdateFlowForResult(
+                    mAppUpdateInfo, AppUpdateType.FLEXIBLE, activity, IN_APP_UPDATE_REQUEST_CODE);
+            if (!success) {
+                Logger.log(TAG, "startUpdate|requested update cannot be started");
+            }
+        } catch (IntentSender.SendIntentException exception) {
+            mInstallStatus = InstallStatus.FAILED;
+            Logger.log(TAG, "startUpdate|failed with : " + exception.getMessage());
+        }
+    }
+
+    @NonNull
+    @Override
+    public AppUpdateState getStatus() {
+        return mAppUpdateState;
+    }
+
+    @Override
+    public void completeUpdate() {
+        mAppUpdateManager.completeUpdate()
+                .addOnSuccessListener(result -> {
+                    Logger.log(TAG, "completeUpdate|was successful");
+                    publishStatus();
+                })
+                .addOnFailureListener(exception -> {
+                    Logger.log(TAG, "completeUpdate|failed with : " + exception.getMessage());
+                    publishStatus();
+                });
+    }
+
+    @Nullable
+    @Override
+    public Integer getProgress() {
+        if (mInstallStatus != InstallStatus.DOWNLOADING) {
+            return null;
+        }
+        return percentageDownloaded;
+    }
+
+    @Override
+    public int getErrorCode() {
+        return mInstallErrorCode;
+    }
+
+    @Override
+    public void skipVersion() {
+        if (mAppUpdateInfo == null) {
+            return;
+        }
+        int currentAvailableVersion = mAppUpdateInfo.availableVersionCode();
+        HiddenPreferences.skipCommCareVersionForUpdate(currentAvailableVersion);
+    }
+
+    //endregion
+
+    @Override
+    public void onStateUpdate(InstallState state) {
+        Logger.log(TAG, "Install state updated with install status: " + state.installStatus()
```

Comment:
since this would get repeatedly called in middle of an update download, we should avoid the general log statement here in order to reduce noise.